test/utils.py

```diff
@@ -0,0 +1,373 @@
+from __future__ import annotations
+
+import asyncio
+import functools
+import multiprocessing as mp
+import multiprocessing.pool
+import multiprocessing.managers
+import os
+import queue
+import time
+import unittest
+from dataclasses import dataclass
+from pathlib import Path
+from typing import overload, TYPE_CHECKING, TypeVar, Protocol, Callable
+from unittest.util import safe_repr
+
+T = TypeVar('T')
+T_contra = TypeVar('T_contra', contravariant=True)
+U = TypeVar('U')
+U_contra = TypeVar('U_contra', contravariant=True)
+
+if TYPE_CHECKING:
+    from _typeshed import SupportsDunderLE, SupportsDunderGE
+
+    class SupportsLeAndGe(SupportsDunderLE[T_contra],
+                          SupportsDunderGE[U_contra],
+                          Protocol[T_contra, U_contra]):
+        pass  # All from inheritance
+
+    class ExtendsAndGE(T, SupportsDunderGE[U], Protocol[T, U]):
+        ...
+
+    class ExtendsAndLE(T, SupportsDunderLE[U], Protocol[T, U]):
+        ...
+
+
+class TestTimeout(Exception):
+    ...
+
+
+def _worker_process(fn, dest, args, kwargs, debug):
+    if debug >= 1:
+        print(f"Worker process running on pid={os.getpid()}")
+    if debug >= 2:
+        time.sleep(15)
+        print('After sleep')
+    # noinspection PyBroadException
+    try:
+        result = fn(*args, *kwargs)
+    except Exception as e:
+        print(e)
+        dest.put(e)
+        raise
+    else:
+        dest.put(result)
+
+
+def run_with_timeout(timeout_sec: int, fn, args, kwargs, debug=0):
+    """fn must be pickleable.
+
+    debug:
+      - 0: normal
+      - 1: print pid on start
+      - 2: also wait 15s in process to allow time to 'Attach to Process' in Pycharm """
+    dest = mp.Queue()
+    # The daemon's main purpose is to stop ourselves accidentally creating a fork-bomb
+    #  by running run_with_timeout in the worker recursively
+    #  (we do weird stuff to get pickle to work which might break?).
+    # This is because daemonic processes cannot spawn more processes (even daemonic ones).
+    # The worker exiting when we get a Ctrl+C is just an added (but necessary) bonus.
+    p = mp.Process(target=_worker_process, args=(fn, dest, args, kwargs, debug), daemon=True)
+    p.start()
+    p.join(timeout_sec)
+    if p.is_alive():
+        p.kill()  # Die!
+        raise TestTimeout(f"Function took too long (exceeded timeout of {timeout_sec}s)")
+    if p.exitcode != 0:
+        raise mp.ProcessError("Error in process") from dest.get()
+    return dest.get()
+
+
+async def join_async(p, timeout: float, interval: float = 0.005, join_timeout: float = 0.0):
```

Comment:
Maybe set the default of `interval` to be 0 - Doing a non-blocking `join` isn't that expensive.
If we do, we need to change the default interval everywhere else as well.